blog/2022-07-18-building-addon-introduction.md

```diff
@@ -0,0 +1,301 @@
+
+```shell
+redis-operator/            # directory name is the same as the addon name
+├── definitions            # stores Definitions, including TraitDefinition, ComponentDefinition, and etc.
+│   └── redis-failover.cue # ComponentDefinition that defines `redis-failover` Components
+├── resources              # resource files, we will refer to them template.cue
+│   ├── crd.yaml           # CRD that comes with Redis Operator (yaml file will be applied directly)
+│   ├── redis-operator.cue # a web-service Component, which installs a Redis Operator
+│   └── topology.cue       # (Optional) to help KubeVela build the relationships of Application resources
+├── metadata.yaml          # addon metadata, including addon name, version, and etc.
+├── parameter.cue          # addon parameters, which users can use to customize their addon installation
+├── README.md              # guides the user to quickly start using your addon
+└── template.cue           # Application description file, which defines an OAM Application
+```
+
+> We will use CUE extensively when writing addons, so [CUE Basic](https://kubevela.io/docs/platform-engineers/cue/basic) might be useful.
+
+### parameter.cue
+
+```cue
+parameter: {
+	//+usage=Redis Operator image.
+	image: *"quay.io/spotahome/redis-operator:v1.1.0" | string
+	// ...omitted
+```
+
+The parameters defined in `parameter.cue` are customizable by users when installing addons, just like Helm Values. You can access the values of these parameters in CUE later by `parameter.<parameter-name>`. In our example, `image` is provided to let the user customize Redis Operator image and can be accessed in `redis-operator.cue` by `parameter.image`.
+
+When you design what parameters are provided to the user, there are several things to consider:
+
+- Do not provide every possible parameter to your user and let them figure out how to dial dozens of knobs by themselves. Abstract fine-grained knobs into some broad parameters, so that the user can input a few parameters and get a usable but customized addon.
+- Decide on sane defaults for starters. Let the user get started with your addon
+  even if they don't provide parameters (if possible).
+- Provide usage for each parameter.
+- Keep parameters consistent across versions to avoid incompatibilities during upgrades.
+
+### template.cue and resources directory
+
+OAM Application is stored here. In our case, we will create an Application that includes a Redis Operator to give our cluster the ability to create Redis clusters.
+
+template.cue and resources directory both serve the same purpose -- to define an Application. Aside from historical reasons, the reason why we split it into two places is readability. When we have a ton of resource in template.cue, it will eventually be too long to read. So we typically put the Application scaffold in `template.cue`, and Application Components in `resources` directory.
```

Comment:
```suggestion
`template.cue` and `resources/` directory both serve the same purpose -- to define an Application. Aside from historical reasons, the reason why we split it into two places is readability. When we have a ton of resource in template.cue, it will eventually be too long to read. So we typically put the Application scaffold in `template.cue`, and Application Components in `resources` directory.
```